modules/javafx.graphics/src/main/java/com/sun/javafx/css/ImmutablePseudoClassSetsCache.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.javafx.css;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+
+/**
+ * A cache for immutable sets of {@link PseudoClass}es.
+ */
+public class ImmutablePseudoClassSetsCache {
+    private static final Map<Set<PseudoClass>, Set<PseudoClass>> CACHE = new HashMap<>();
+
+    /**
+     * Returns an immutable set of {@link PseudoClass}es.
+     * <p>
+     * Note: this method may or may not return the same instance for the same set of
+     * {@link PseudoClass}es.
+     *
+     * @param pseudoClasses a set of {@link PseudoClass} to make immutable, cannot be {@code null}
+     * @return an immutable set of {@link PseudoClass}es, never {@code null}
+     * @throws NullPointerException when {@code pseudoClasses} is {@code null} or contains {@code null}s
+     */
+    public static Set<PseudoClass> of(Set<PseudoClass> pseudoClasses) {
+        return CACHE.computeIfAbsent(
+            Objects.requireNonNull(pseudoClasses, "pseudoClasses cannot be null"),
+            k -> Set.copyOf(pseudoClasses)
```

Comment:
The call to `computeIfAbsent` will always do the following things:
1. Capture the `pseudoClasses` argument
2. Call `Map.get` internally
3. ...which calls `Set.hashCode`
4. ...which, for `AbstractSet.hashCode`, iterates over the entire set using an iterator and calls `hashCode` for each element.

Since there is no fast path, I wonder whether this may have negative performance implications for larger scene graphs.

Reply:
I don't think there is much of an impact, scene graph size should not affect it.  The sets involved are tiny, usually consisting of 1 or 2 items and very rarely 3 or possibly more(*). The hash code of `PseudoClass` is very fast, as they're singletons which don't override their hash code.

(*) The sets are based on pseudo class combinations encountered in a Scene that would have an active role in changing the control's visual appearance.  A pseudo class that doesn't affect appearance is filtered.  This means usually only states like `focused` or `armed` have an affect.  A class like `focused-within` would only be part of the set if the control involved has styles based on that class that would affect is appearance.

Reply:
Couldn't you just add a fast path by calculating the hash code of the immutable set eagerly and storing it in a field? This entire method could probably be written to be allocation-free if the set passed into the method is already an immutable set.

Reply:
I think I am missing something :)

The set that is passed in may indeed be immutable, but it may not be the set we have cached.  The memory savings only work when I then return the cached set; if I return the set you pass in (even if it is immutable) then we would still have large numbers of `Set<PseudoClass>` that are duplicates.

So, at a minimum, the `Set` that is passed in would need its hash code calculated so I can check if it is the cached version, or otherwise return the cached version.

I could cache the hash code for the sets that are returned here (they are immutable sets returned by `Set.copyOf` -- I checked their code, and they don't cache the hash codes).  That would however only help if you often pass in a set that is already cached, to find out if you got the same one.  Perhaps this happens a lot, I could test that.

Note that `HashMap` already caches hash codes for things that are part of the map, so only the hash code of the input needs to be calculated on this call.

Reply:
About allocation free: if the `Set` you pass in is immutable, and it is has a cached equivalent (or it is the same one), then no allocations occur (calculating the hash code does not do allocations, unless someone overrides hash code to do so) -- `copyOf` will not allocate anything if the `Set` was immutable already.

Allocations only occur in two cases:
- When the set you pass in is not immutable, and it is a new set that wasn't cached yet.  A copy is made (allocation) and a `Map.Entry` is created (allocation)
- When the set you pass in is mutable, but wasn't cached yet.  No copy is needed, but still a `Map.Entry` is created.

These allocations however will be only during application startup (when new combinations of pseudo classes are encountered) and the first time some state occurs in your application that results in a new pseudo class combination.  The amount of allocations over the life time of your application (from this code) is probably around 50-100.

Reply:
Perhaps write some code to show what you mean here?

Reply:
> I could cache the hash code for the sets that are returned here (they are immutable sets returned by `Set.copyOf` -- I checked their code, and they don't cache the hash codes). That would however only help if you often pass in a set that is already cached, to find out if you got the same one. Perhaps this happens a lot, I could test that.

That's what I meant to say. It seems like an easy win to cache the hash code of immutable sets if those same immutable sets can concievably be passed into this method quite often.

Reply:
> Allocations only occur in two cases:
> 
> * When the set you pass in is not immutable, and it is a new set that wasn't cached yet.  A copy is made (allocation) and a `Map.Entry` is created (allocation)
> * When the set you pass in is immutable, but wasn't cached yet.  No copy is needed, but still a `Map.Entry` is created.

This method will allocate on every single call because it captures the method argument in a lambda object (but it doesn't need to, it can be written in a way that doesn't use capturing lambdas). That's another easy win to be had if this method is called often.